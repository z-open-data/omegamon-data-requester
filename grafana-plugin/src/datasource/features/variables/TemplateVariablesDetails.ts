import { TypedVariableModel, ScopedVars, VariableType, CustomVariableModel, QueryVariableModel } from '@grafana/data';
import { VariableInterpolation, getTemplateSrv } from '@grafana/runtime';
import { normalizeToArray } from 'public-common';

import { getVariablesFromStr } from 'datasource/features/cmsSqlTransformers/cmsSqlToMetricsQuery/tokenizer';

import { emptyStringFiller } from './emptyStringFiller';

const ALL_VAR_VALUE = '$__all';

type VariableInfo = {
  interpolation: VariableInterpolation;
  model: TypedVariableModel;
  /** The Grafana's interpretation of an empty string as custom "All" value is not explicitly defined. */
  customAllValue?: string;
  isAllSelected: boolean;
};

/** This class encapsulates details about a variable template acquired from the Grafana. */
export class TemplateVariablesDetails {
  private template: string;
  /** An object containing Grafana's environments scoped variables. */
  private scopedVars: ScopedVars;
  /** An array of variable information objects that are used within the template. Please don't use it. */
  private usedVariables: VariableInfo[];

  /**
   * Throws if template contains variables that are not found in dashboard
   */
  constructor(template: string, scopedVars: ScopedVars) {
    this.template = template;
    this.scopedVars = scopedVars;
    this.usedVariables = [];
    if (!getVariablesFromStr(template, 0).length) {
      return;
    }
    const interpolations = this.getVariableInterpolations();
    this.usedVariables = interpolations.map((interpolation) => {
      const model = TemplateVariablesDetails.getVariableModel(interpolation.variableName);
      if (!model) {
        throw new Error(`There is no variable information for "${interpolation.match}"`);
      }

      const isAllSelected = TemplateVariablesDetails.isCurrentValueAll(model);
      const customAllValue = isAllSelected && model.allValue ? model.allValue : undefined;

      return { interpolation, model, customAllValue, isAllSelected };
    });
  }

  containsVariables() {
    return this.usedVariables.length !== 0;
  }

  getVariations(): [string, ...string[]] {
    return this.applyVariations(this.template, this.usedVariables);
  }

  private applyVariations(template: string, variablesInfo: VariableInfo[]): [string, ...string[]] {
    const [firstVariableInfo, ...restVariablesInfo] = variablesInfo;
    if (!firstVariableInfo) {
      return [template];
    }

    const { match } = firstVariableInfo.interpolation;
    const variableValues = TemplateVariablesDetails.getCurrentValues(firstVariableInfo);
    const resultVariations: string[] = variableValues.flatMap((value) => {
      const filledTemplate = template.replace(match, value);
      return this.applyVariations(filledTemplate, restVariablesInfo);
    });
    return resultVariations as [string, ...string[]];
  }

  /** Returns true if template contains only single variable and nothing else (not even literals)
   * '$Variable' -> true
   * '${Variable}' -> true
   * '${Variable}test' -> false
   * '${Variable}${Variable}' -> false
   */
  isSingleVariableExpression(): boolean {
    return this.usedVariables[0]?.interpolation.match === this.template;
  }

  /** This method verifies whether any variable within the "usedVariables" has "All" selected without a custom value specified */
  hasUnspecifiedAllSelected(): boolean {
    /* The Grafana's interpretation of an empty string as custom "All" value is not explicitly defined.
     * Therefore we handle both empty strings and undefined values in the same manner. */
    return this.usedVariables.some((usedVariable) => usedVariable.isAllSelected && !usedVariable.customAllValue);
  }

  private static getCurrentValues(variableInfo: VariableInfo): string[] {
    if (!TemplateVariablesDetails.variableHasType(variableInfo, ['query', 'custom', 'constant', 'textbox'] as const)) {
      throw new Error(
        `Variable "${variableInfo.interpolation.match}" with type "${variableInfo.model.type}" is not supported in this context`
      );
    }
    const { model } = variableInfo;
    if (variableInfo.isAllSelected) {
      /* The Grafana's interpretation of an empty string as custom "All" value is not explicitly defined.
       * Therefore we handle both empty strings and undefined values in the same manner. */
      if (variableInfo.customAllValue) {
        return [variableInfo.customAllValue];
      }
      // In Grafana, option value can be represented as either a single string or an array of strings, which is why we utilize 'flatMap'.
      return model.options.filter((option) => option.value !== ALL_VAR_VALUE).flatMap((option) => option.value);
    }

    // We turned empty strings into <empty value> in query variable values
    const fixedValues = normalizeToArray(model.current.value).map((value) =>
      value === emptyStringFiller ? '' : value
    );

    return fixedValues;
  }

  private static getVariableModel(variableName: string): TypedVariableModel | null {
    const variablesModels = getTemplateSrv().getVariables();
    const variableModel = variablesModels.find((variable) => variable.name === variableName);
    return variableModel ?? null;
  }

  private getVariableInterpolations(): VariableInterpolation[] {
    const variableInterpolations: VariableInterpolation[] = [];
    getTemplateSrv().replace(this.template, this.scopedVars, '', variableInterpolations);
    return variableInterpolations;
  }

  private static variableHasType<T extends readonly VariableType[]>(
    variableInfo: VariableInfo,
    types: T
  ): variableInfo is VariableInfo & { model: { type: T[number] } } {
    return types.includes(variableInfo.model.type);
  }

  private static isCurrentValueAll(
    varInfo: TypedVariableModel | null
  ): varInfo is CustomVariableModel | QueryVariableModel {
    if (!TemplateVariablesDetails.variableSupportsAll(varInfo)) {
      return false;
    }
    //  Value is array for multi variable (multi option is true)
    return varInfo.current.value[0] === '$__all' || varInfo.current.value === '$__all';
  }

  private static variableSupportsAll(
    varInfo: TypedVariableModel | null | undefined
  ): varInfo is CustomVariableModel | QueryVariableModel {
    const VARIABLE_TYPES_WITH_ALL_SUPPORT: VariableType[] = ['query', 'custom'];
    return VARIABLE_TYPES_WITH_ALL_SUPPORT.includes(varInfo?.type as VariableType);
  }
}
