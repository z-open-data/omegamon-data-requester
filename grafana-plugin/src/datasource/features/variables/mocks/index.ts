export * from './metricsQueryFilters';
export * from './mockedTableMetadatas';
export * from './replaceFnInputOutputMap';
export * from './scopedVars';
export * from './variableModels';
