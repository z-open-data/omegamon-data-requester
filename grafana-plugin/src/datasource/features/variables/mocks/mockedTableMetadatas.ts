import { TableMetadata, AffinityId } from 'public-domain';

export const mockedTableMetadata: TableMetadata = {
  id: 'CICSCAD',
  name: 'CICSplex Association Data Details',
  description:
    'The CICSplex Association Data Details attribute group allows you to display the association data related to an individual transaction.',
  productCode: 'CP',
  applicationCode: 'KCP',
  applicationName: 'CICS',
  sampleType: 'pure',
  isMultiRow: true,
  columns: {
    VLUNAME: {
      id: 'VLUNAME',
      name: 'Originating System ID',
      attributeName: 'QQQQ',
      description: 'Is the combination of z/OS System ID (SMFID) and CICS region name.',
      type: 'string',
      maxLength: 123123,
      scaleFactor: 123123,
      sortType: 'lexical',
      defaultSortDirection: 'A',
      version: 0,
      extensions: {},
      atomize: false,
    },
    DB2SUBSYSTEM: {
      id: 'DB2SUBSYSTEM',
      name: 'Originating System ID',
      attributeName: 'QQQQ',
      description: 'Is the combination of z/OS System ID (SMFID) and CICS region name.',
      type: 'string',
      maxLength: 123123,
      scaleFactor: 123123,
      sortType: 'lexical',
      defaultSortDirection: 'A',
      version: 0,
      extensions: {},
      atomize: false,
    },
    JOBNAME: {
      id: 'JOBNAME',
      name: 'Originating System ID',
      attributeName: 'QQQQ',
      description: 'Is the combination of z/OS System ID (SMFID) and CICS region name.',
      type: 'string',
      maxLength: 123123,
      scaleFactor: 123123,
      sortType: 'lexical',
      defaultSortDirection: 'A',
      version: 0,
      extensions: {},
      atomize: false,
    },
    SVCCLASS: {
      id: 'SVCCLASS',
      name: 'Originating System ID',
      attributeName: 'QQQQ',
      description: 'Is the combination of z/OS System ID (SMFID) and CICS region name.',
      type: 'string',
      maxLength: 123123,
      scaleFactor: 123123,
      sortType: 'lexical',
      defaultSortDirection: 'A',
      version: 0,
      extensions: {},
      atomize: false,
    },
    DB2ID: {
      id: 'DB2ID',
      name: 'Originating System ID',
      attributeName: 'QQQQ',
      description: 'Is the combination of z/OS System ID (SMFID) and CICS region name.',
      type: 'string',
      maxLength: 123123,
      scaleFactor: 123123,
      sortType: 'lexical',
      defaultSortDirection: 'A',
      version: 0,
      extensions: {},
      atomize: false,
    },
    STEPNAME: {
      id: 'STEPNAME',
      name: 'Originating System ID',
      attributeName: 'QQQQ',
      description: 'Is the combination of z/OS System ID (SMFID) and CICS region name.',
      type: 'string',
      maxLength: 123123,
      scaleFactor: 123123,
      sortType: 'lexical',
      defaultSortDirection: 'A',
      version: 0,
      extensions: {},
      atomize: false,
    },
    MGTSTAT: {
      id: 'MGTSTAT',
      name: 'Management Status',
      attributeName: 'Management_Status',
      odpName: '',
      description: '',
      type: 'number',
      maxLength: 0,
      sortType: 'lexical',
      version: 1,
      defaultSortDirection: 'A',
      enums: {
        '0000': 'Unknown',
        '0001': 'Monitored',
        '0002': 'Managed',
        '0003': 'Isolated',
        '0004': 'NonSwap',
        '0005': 'OK',
        '0X00': 'Unknown',
        '0X01': 'Monitored',
        '0X02': 'Managed',
        '0X03': 'Isolated',
        '0X04': 'NonSwap',
        '0X05': 'OK',
      },
      extensions: {},
      scaleFactor: 0,
      precision: 0,
      printf: '',
      atomize: false,
    },
    SWPSTAT: {
      id: 'SWPSTAT',
      name: 'Swap Status',
      attributeName: 'Swap_Status',
      odpName: '',
      description: '',
      type: 'number',
      maxLength: 0,
      sortType: 'lexical',
      version: 1,
      defaultSortDirection: 'A',
      enums: {
        '0000': 'Unknown',
        '0001': 'In',
        '0002': 'Out',
        '0003': 'InNSW',
        '0004': 'OutLSW',
        '0X00': 'Unknown',
        '0X01': 'In',
        '0X02': 'Out',
        '0X03': 'InNSW',
        '0X04': 'OutLSW',
      },
      extensions: {},
      scaleFactor: 0,
      precision: 0,
      printf: '',
      atomize: false,
    },
    FRMSUSEPCT: {
      id: 'FRMSUSEPCT',
      name: 'Frames Used Percent',
      attributeName: 'Frames_Used_Percent',
      odpName: '',
      description: '',
      type: 'number',
      maxLength: 0,
      sortType: 'lexical',
      version: 6,
      defaultSortDirection: 'A',
      enums: {},
      extensions: {},
      scaleFactor: 1,
      precision: 4,
      printf: '',
      atomize: false,
    },
  },
  affinityEntity: {
    id: '%IBM.STATIC011' as AffinityId,
    displayText: 'MVS CICS',
    productCode: 'KCP',
    symbol: 'MVS_CICS',
  },
  version: 4,
};
