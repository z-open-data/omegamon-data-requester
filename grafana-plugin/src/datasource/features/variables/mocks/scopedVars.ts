import { ScopedVars } from '@grafana/data';

export const scopedVars = {
  __interval: {
    text: '30s',
    value: '30s',
  },
  __interval_ms: {
    text: '30000',
    value: 30000,
  },
} satisfies ScopedVars;
