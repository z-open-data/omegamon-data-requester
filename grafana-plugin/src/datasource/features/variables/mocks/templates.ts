export const templateMocks = {
  jobname: '${JOBNAME}',
  stepname: '$STEPNAME',
  svcclass: '${SVCCLASS}',
  db2id: '${DB2ID}',
  vluname: '${VLUNAME}',
  db2subsystem: '${DB2SUBSYSTEM}',
  invalid: '${VLUNAME}test${VLUNAME}',
  multiVariables: 'foo${JOBNAME}bar${SVCCLASS}baz$STEPNAME',
  interval: '${INTERVAL}',
  vlunameDb2subsystem: '${VLUNAME}${DB2SUBSYSTEM}',
  originnode: '$ORIGINNODE',
  mvsid: '${MVSID}',
  mvsidWithLiteral: '${MVSID}test',
  mvsidTwoVariables: '${MVSID}${MVSID}',
  twoVariable: '${MVSID}${SVCCLASS}',
  cicsid: '${CICSID}',
  mgtstat: '$MGTSTAT',
  swpstat: '${SWPSTAT}',
  frmsusepct: '$FRMSUSEPCT',
};
