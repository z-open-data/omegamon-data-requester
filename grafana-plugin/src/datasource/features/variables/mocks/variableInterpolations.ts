import { VariableInterpolation } from '@grafana/runtime';

export const variableInterpolationMocks = {
  notFound: {
    match: '$VarNotFoundInDashboard',
    variableName: 'VarNotFoundInDashboard',
    format: '',
    value: '$VarNotFoundInDashboard',
    found: false,
  },
  vluname: {
    match: '${VLUNAME}',
    variableName: 'VLUNAME',
    format: '',
    value: 'OBC1,OBC2',
    found: true,
  },
  jobname: {
    match: '${JOBNAME}',
    variableName: 'JOBNAME',
    format: '',
    value: 'RMFGAT',
    found: true,
  },
  stepname: {
    match: '$STEPNAME',
    variableName: 'STEPNAME',
    format: '',
    value: 'XCFAS,WLM,RMFGAT',
    found: true,
  },
  svcclass: {
    match: '${SVCCLASS}',
    variableName: 'SVCCLASS',
    format: '',
    value: 'USSDSYS,IMSRGNS,SYSSTC',
    found: true,
  },
  db2subsystem: {
    match: '${DB2SUBSYSTEM}',
    variableName: 'DB2SUBSYSTEM',
    format: '',
    value: 'OBC1,OBC2',
    found: true,
  },
  db2id: {
    match: '${DB2ID}',
    variableName: 'DB2ID',
    format: '',
    value: 'OBC4',
    found: true,
  },
  mvsid: {
    match: '${MVSID}',
    variableName: 'MVSID',
    format: '',
    value: 'RS25',
    found: true,
  },
  interval: {
    match: '${INTERVAL}',
    variableName: 'INTERVAL',
    format: '',
    value: '',
    found: true,
  },
  originnode: {
    match: '$ORIGINNODE',
    variableName: 'ORIGINNODE',
    format: '',
    value: 'RSPLUI01:RS25:MVSSYS',
    found: true,
  },
  cicsid: {
    match: '${CICSID}',
    variableName: 'CICSID',
    format: '',
    value: '',
    found: true,
  },
  mgtstat: {
    match: '$MGTSTAT',
    variableName: 'MGTSTAT',
    format: '',
    value: 'Unknown',
    found: true,
  },
  swpstat: {
    match: '${SWPSTAT}',
    variableName: 'SWPSTAT',
    format: '',
    value: 'OutLSW',
    found: true,
  },
  frmsusepct: {
    match: '$FRMSUSEPCT',
    variableName: 'FRMSUSEPCT',
    format: '',
    value: '43.4',
    found: true,
  },
} satisfies Record<string, VariableInterpolation>;
