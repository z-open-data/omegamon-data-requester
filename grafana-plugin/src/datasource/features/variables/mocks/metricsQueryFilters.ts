import { MetricsQueryFilter, MetricsQueryFilters, MetricsQueryFilterClause } from 'datasource/domain';

import { templateMocks } from './templates';

export const andFilter = [
  {
    clause: {
      columnId: 'SVCCLASS',
      operator: '=',
      userDefinedValue: templateMocks.svcclass,
    },
  },
  {
    clause: {
      columnId: 'STEPNAME',
      operator: '=',
      userDefinedValue: templateMocks.stepname,
    },
  },
] satisfies MetricsQueryFilter[];

export const compoundFilter = {
  nonAggregated: {
    or: [
      {
        clause: {
          columnId: 'JOBNAME',
          operator: '=',
          userDefinedValue: templateMocks.jobname,
        },
      },
      {
        and: andFilter,
      },
    ],
  },
} satisfies MetricsQueryFilters;

export const multiVariablesTemplateFilter = {
  nonAggregated: {
    clause: {
      columnId: 'JOBNAME',
      operator: '=',
      userDefinedValue: templateMocks.multiVariables,
    },
  },
} satisfies MetricsQueryFilters;

export const varNotFoundInDashboardFilter = {
  nonAggregated: {
    clause: {
      columnId: 'JOBNAME',
      operator: '=',
      userDefinedValue: '$VarNotFoundInDashboard',
    },
  },
} satisfies MetricsQueryFilters;

export const singleVarFilter = {
  clause: {
    columnId: 'DB2ID',
    operator: '=',
    userDefinedValue: '${DB2ID}',
  },
} satisfies MetricsQueryFilter;

export const noVarInTemplateFilter = {
  nonAggregated: {
    clause: {
      columnId: 'DB2ID',
      operator: '=',
      userDefinedValue: 'OBC4',
    },
  },
} satisfies MetricsQueryFilters;

export const multiValueClause = {
  columnId: 'SVCCLASS',
  operator: '=',
  userDefinedValue: '${SVCCLASS}',
} satisfies MetricsQueryFilterClause;

export const allValFilter = {
  nonAggregated: {
    clause: {
      columnId: 'VLUNAME',
      operator: '=',
      userDefinedValue: '${VLUNAME}',
    },
  },
} satisfies MetricsQueryFilters;

export const invalidTemplateAllFilter = {
  nonAggregated: {
    clause: {
      columnId: 'VLUNAME',
      operator: '=',
      userDefinedValue: '${VLUNAME}test${VLUNAME}',
    },
  },
} satisfies MetricsQueryFilters;

export const allInAndFilter = {
  nonAggregated: {
    and: [
      {
        clause: {
          columnId: 'VLUNAME',
          operator: '=',
          userDefinedValue: '${VLUNAME}',
        },
      },
      {
        clause: {
          columnId: 'DB2ID',
          operator: '=',
          userDefinedValue: '${DB2ID}',
        },
      },
    ],
  },
} satisfies MetricsQueryFilters;

export const allInMultipleClauseAndFilter = {
  nonAggregated: {
    and: [
      {
        clause: {
          columnId: 'VLUNAME',
          operator: '=',
          userDefinedValue: '${VLUNAME}',
        },
      },
      {
        clause: {
          columnId: 'DB2ID',
          operator: '=',
          userDefinedValue: '${DB2ID}',
        },
      },
      {
        clause: {
          columnId: 'SVCCLASS',
          operator: '=',
          userDefinedValue: 'USSDSYS',
        },
      },
    ],
  },
} satisfies MetricsQueryFilters;

export const allInOrFilter = {
  nonAggregated: {
    or: [
      {
        clause: {
          columnId: 'VLUNAME',
          operator: '=',
          userDefinedValue: '${VLUNAME}',
        },
      },
      {
        clause: {
          columnId: 'DB2ID',
          operator: '=',
          userDefinedValue: '${DB2ID}',
        },
      },
      {
        clause: {
          columnId: 'SVCCLASS',
          operator: '=',
          userDefinedValue: 'USSDSYS',
        },
      },
    ],
  },
} satisfies MetricsQueryFilters;

export const complexAllFilter = {
  nonAggregated: {
    and: [
      {
        clause: {
          columnId: 'JOBNAME',
          operator: '=',
          userDefinedValue: '${JOBNAME}',
        },
      },
      {
        or: [
          {
            clause: {
              columnId: 'VLUNAME',
              operator: '=',
              userDefinedValue: '${VLUNAME}',
            },
          },
          {
            clause: {
              columnId: 'DB2ID',
              operator: '=',
              userDefinedValue: '${DB2ID}',
            },
          },
          {
            clause: {
              columnId: 'SVCCLASS',
              operator: '=',
              userDefinedValue: 'USSDSYS',
            },
          },
        ],
      },
    ],
  },
} satisfies MetricsQueryFilters;

export const customAllFilter = {
  nonAggregated: {
    and: [
      {
        clause: {
          columnId: 'VLUNAME',
          operator: '=',
          userDefinedValue: '${VLUNAME}',
        },
      },
      {
        or: [
          {
            clause: {
              columnId: 'DB2SUBSYSTEM',
              operator: '=',
              userDefinedValue: '${DB2SUBSYSTEM}',
            },
          },
          {
            clause: {
              columnId: 'DB2ID',
              operator: '=',
              userDefinedValue: '${DB2ID}',
            },
          },
          {
            clause: {
              columnId: 'SVCCLASS',
              operator: '=',
              userDefinedValue: 'USSDSYS',
            },
          },
        ],
      },
    ],
  },
} satisfies MetricsQueryFilters;

export const mgtstatClause = {
  columnId: 'MGTSTAT',
  operator: '=',
  userDefinedValue: '$MGTSTAT',
} satisfies MetricsQueryFilterClause;

export const swpstatClause = {
  columnId: 'SWPSTAT',
  operator: '=',
  userDefinedValue: '${SWPSTAT}',
} satisfies MetricsQueryFilterClause;

export const frmsusepctClause = {
  columnId: 'FRMSUSEPCT',
  operator: '=',
  userDefinedValue: '$FRMSUSEPCT',
} satisfies MetricsQueryFilterClause;

export const mgtstatFilter = {
  nonAggregated: {
    clause: mgtstatClause,
  },
} satisfies MetricsQueryFilters;

export const swpstatFilter = {
  nonAggregated: {
    clause: swpstatClause,
  },
} satisfies MetricsQueryFilters;

export const frmsusepctFilter = {
  nonAggregated: {
    clause: frmsusepctClause,
  },
} satisfies MetricsQueryFilters;
