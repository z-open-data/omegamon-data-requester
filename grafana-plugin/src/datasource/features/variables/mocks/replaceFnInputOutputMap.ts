import { VariableInterpolation } from '@grafana/runtime';

import { varNotFoundInDashboardFilter } from './metricsQueryFilters';
import { templateMocks } from './templates';
import { variableInterpolationMocks } from './variableInterpolations';

type MockedReplaceFnInputOutputMap = {
  [template: string]: VariableInterpolation[];
};

export const mockedReplaceFnInputOutputMap = {
  [templateMocks.invalid]: [variableInterpolationMocks.vluname, variableInterpolationMocks.vluname],

  [templateMocks.jobname]: [variableInterpolationMocks.jobname],
  [templateMocks.stepname]: [variableInterpolationMocks.stepname],
  [templateMocks.svcclass]: [variableInterpolationMocks.svcclass],
  [templateMocks.db2id]: [variableInterpolationMocks.db2id],
  [templateMocks.vluname]: [variableInterpolationMocks.vluname],
  [templateMocks.db2subsystem]: [variableInterpolationMocks.db2subsystem],
  [templateMocks.multiVariables]: [
    variableInterpolationMocks.jobname,
    variableInterpolationMocks.svcclass,
    variableInterpolationMocks.stepname,
  ],
  [varNotFoundInDashboardFilter.nonAggregated.clause.userDefinedValue]: [variableInterpolationMocks.notFound],
  [templateMocks.interval]: [variableInterpolationMocks.interval],
  [templateMocks.vlunameDb2subsystem]: [variableInterpolationMocks.vluname, variableInterpolationMocks.db2subsystem],
  [templateMocks.originnode]: [variableInterpolationMocks.originnode],
  [templateMocks.mvsid]: [variableInterpolationMocks.mvsid],
  [templateMocks.mvsidWithLiteral]: [variableInterpolationMocks.mvsid],
  [templateMocks.mvsidTwoVariables]: [variableInterpolationMocks.mvsid, variableInterpolationMocks.mvsid],
  [templateMocks.twoVariable]: [variableInterpolationMocks.mvsid, variableInterpolationMocks.svcclass],
  [templateMocks.cicsid]: [variableInterpolationMocks.cicsid],
  [templateMocks.mgtstat]: [variableInterpolationMocks.mgtstat],
  [templateMocks.swpstat]: [variableInterpolationMocks.swpstat],
  [templateMocks.frmsusepct]: [variableInterpolationMocks.frmsusepct],
} satisfies MockedReplaceFnInputOutputMap;
