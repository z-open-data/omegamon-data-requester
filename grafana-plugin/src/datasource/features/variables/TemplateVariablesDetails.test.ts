import { ScopedVars } from '@grafana/data';
import { VariableInterpolation } from '@grafana/runtime';

import { mockedReplaceFnInputOutputMap, scopedVars, mockedTypedVariableModels } from './mocks';
import { TemplateVariablesDetails } from './TemplateVariablesDetails';

jest.mock('@grafana/runtime', () => ({
  getTemplateSrv: () => ({
    replace: (template: string, scopedVars: ScopedVars, format: string, interpolations: VariableInterpolation[]) => {
      const foundInterpolations = mockedReplaceFnInputOutputMap[template];
      interpolations.push(...(foundInterpolations || []));
      return template;
    },
    getVariables: () => {
      return mockedTypedVariableModels;
    },
  }),
}));

describe('getVariations', () => {
  it('should return single value for variable that is non-multivalue', () => {
    const singleValueVariableTemplateDetails = new TemplateVariablesDetails('$ORIGINNODE', scopedVars);
    expect(singleValueVariableTemplateDetails.getVariations()).toMatchSnapshot();
  });
  it('should return single value for variable that is multi-value and has single value selected', () => {
    const singleValueSelectedTemplateDetails = new TemplateVariablesDetails('${MVSID}', scopedVars);
    expect(singleValueSelectedTemplateDetails.getVariations()).toMatchSnapshot();
  });
  it('should return empty array for variable that is multi-value but no value is selected', () => {
    const noValueSelectedTemplateDetails = new TemplateVariablesDetails('${CICSID}', scopedVars);
    expect(noValueSelectedTemplateDetails.getVariations()).toEqual([]);
  });
  it('should return all possible strings variations for template with multiple variables', () => {
    const multiVariableTemplateDetails = new TemplateVariablesDetails(
      'foo${JOBNAME}bar${SVCCLASS}baz$STEPNAME',
      scopedVars
    );
    expect(multiVariableTemplateDetails.getVariations()).toMatchSnapshot();
  });
  it('should return the provided custom "All" value when it is specified', () => {
    const specifiedAllTemplateDetails = new TemplateVariablesDetails('${DB2SUBSYSTEM}', scopedVars);
    expect(specifiedAllTemplateDetails.getVariations()).toMatchSnapshot();
  });
  it('should return all possible values for the variable when unspecified "All" is selected', () => {
    const unspecifiedAllTemplateDetails = new TemplateVariablesDetails('${VLUNAME}', scopedVars);
    expect(unspecifiedAllTemplateDetails.getVariations()).toMatchSnapshot();
  });
  it('should return the received string unchanged when the string does not contain a variable', () => {
    const noVariableTemplateDetails = new TemplateVariablesDetails('JOBNAME', scopedVars);
    expect(noVariableTemplateDetails.getVariations()).toMatchSnapshot();
  });
  it('should throw error when the template that uses wrong variable type is received', () => {
    const intervalVariablesTemplateDetails = new TemplateVariablesDetails('${INTERVAL}', scopedVars);
    expect(() => intervalVariablesTemplateDetails.getVariations()).toThrowError(
      'Variable "${INTERVAL}" with type "interval" is not supported in this context'
    );
  });
  it('should throw error when variable is not found in dashboard', () => {
    expect(() => {
      const nonExistingVariablesTemplateDetails = new TemplateVariablesDetails('$VarNotFoundInDashboard', scopedVars);
      nonExistingVariablesTemplateDetails.getVariations();
    }).toThrowError('There is no variable information for "$VarNotFoundInDashboard"');
  });
});

describe('hasUnspecifiedAllSelected', () => {
  it('should return false when variable has specified custom "All"', () => {
    const specifiedAllTemplateDetails = new TemplateVariablesDetails('${DB2SUBSYSTEM}', scopedVars);
    expect(specifiedAllTemplateDetails.hasUnspecifiedAllSelected()).toBe(false);
  });
  it('should return false when none of template variables has "All" value', () => {
    const multiVariableTemplateDetails = new TemplateVariablesDetails(
      'foo${JOBNAME}bar${SVCCLASS}baz$STEPNAME',
      scopedVars
    );
    expect(multiVariableTemplateDetails.hasUnspecifiedAllSelected()).toBe(false);
  });
  it('should return true when unspecified "All" is selected in one of the few template variables', () => {
    const unspecifiedAllTemplateDetails = new TemplateVariablesDetails('${VLUNAME}${DB2SUBSYSTEM}', scopedVars);
    expect(unspecifiedAllTemplateDetails.hasUnspecifiedAllSelected()).toBe(true);
  });
  it('should return true when template has single variable with unspecified "All" value', () => {
    const singleUnspecifiedAllTemplateDetails = new TemplateVariablesDetails('${VLUNAME}', scopedVars);
    expect(singleUnspecifiedAllTemplateDetails.hasUnspecifiedAllSelected()).toBe(true);
  });
});

describe('isSingleVariableExpression', () => {
  it('should return true when provided template is single variable expression, so it is equal to first variable match', () => {
    const singleExpressionVariablesDetails = new TemplateVariablesDetails('${DB2SUBSYSTEM}', scopedVars);
    expect(singleExpressionVariablesDetails.isSingleVariableExpression()).toEqual(true);
  });
  it('should return false when provided template is not a single variable expression and is not equal to first found variable match', () => {
    const multiVariableTemplateDetails = new TemplateVariablesDetails(
      'foo${JOBNAME}bar${SVCCLASS}baz$STEPNAME',
      scopedVars
    );
    expect(multiVariableTemplateDetails.isSingleVariableExpression()).toEqual(false);
  });
  it('should return false when provided template is empty', () => {
    const emptyTemplateVariablesDetails = new TemplateVariablesDetails('', scopedVars);
    expect(emptyTemplateVariablesDetails.isSingleVariableExpression()).toEqual(false);
  });
  it('should return false when provided template does not contain a variable', () => {
    const noVariableTemplateVariablesDetails = new TemplateVariablesDetails('NoVarsTemplate', scopedVars);
    expect(noVariableTemplateVariablesDetails.isSingleVariableExpression()).toEqual(false);
  });
  it('should return false when provided template contains variable syntax and literal', () => {
    const mixedSyntaxTemplateVariablesDetails = new TemplateVariablesDetails('${MVSID}test', scopedVars);
    expect(mixedSyntaxTemplateVariablesDetails.isSingleVariableExpression()).toEqual(false);
  });
  it('should return false when provided template contains two variables', () => {
    const twoVariablesTemplateVariablesDetails = new TemplateVariablesDetails('${MVSID}${SVCCLASS}', scopedVars);
    expect(twoVariablesTemplateVariablesDetails.isSingleVariableExpression()).toEqual(false);
  });
  it('should return false when provided template contains two variables that are the same', () => {
    const twoSameVariablesTemplateVariablesDetails = new TemplateVariablesDetails('${MVSID}${MVSID}', scopedVars);
    expect(twoSameVariablesTemplateVariablesDetails.isSingleVariableExpression()).toEqual(false);
  });
});
