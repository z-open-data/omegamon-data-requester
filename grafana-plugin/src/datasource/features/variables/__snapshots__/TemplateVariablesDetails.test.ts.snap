// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getVariations should return all possible strings variations for template with multiple variables 1`] = `
[
  "fooRMFGATbarUSSDSYSbazXCFAS",
  "fooRMFGATbarUSSDSYSbazWLM",
  "fooRMFGATbarUSSDSYSbazRMFGAT",
  "fooRMFGATbarIMSRGNSbazXCFAS",
  "fooRMFGATbarIMSRGNSbazWLM",
  "fooRMFGATbarIMSRGNSbazRMFGAT",
  "fooRMFGATbarSYSSTCbazXCFAS",
  "fooRMFGATbarSYSSTCbazWLM",
  "fooRMFGATbarSYSSTCbazRMFGAT",
]
`;

exports[`getVariations should return all possible values for the variable when unspecified "All" is selected 1`] = `
[
  "OBC1",
  "OBC2",
]
`;

exports[`getVariations should return single value for variable that is multi-value and has single value selected 1`] = `
[
  "RS25",
]
`;

exports[`getVariations should return single value for variable that is non-multivalue 1`] = `
[
  "RSPLUI01:RS25:MVSSYS",
]
`;

exports[`getVariations should return the provided custom "All" value when it is specified 1`] = `
[
  "OCA5",
]
`;

exports[`getVariations should return the received string unchanged when the string does not contain a variable 1`] = `
[
  "JOBNAME",
]
`;
